In [4]:

# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install statsmodels linearmodels
#!{sys.executable} -m pip install --upgrade pip

    100% |████████████████████████████████| 942kB 2.6MB/s 


https://github.com/bashtage/linearmodels

In [5]:
import numpy as np
from statsmodels.datasets import grunfeld
data = grunfeld.load_pandas().data
data.year = data.year.astype(np.int64)
# MultiIndex, entity - time
data = data.set_index(['firm','year'])
from linearmodels import PanelOLS
mod = PanelOLS(data.invest, data[['value','capital']], entity_effects=True)
res = mod.fit(cov_type='clustered', cluster_entity=True)

In [7]:
data

invest     value   capital
firm              year                              
b'General Motors' 1935   317.600  3078.500     2.800
                  1936   391.800  4661.700    52.600
                  1937   410.600  5387.100   156.900
                  1938   257.700  2792.200   209.200
                  1939   330.800  4313.200   203.400
                  1940   461.200  4643.900   207.200
                  1941   512.000  4551.200   255.200
                  1942   448.000  3244.100   303.700
                  1943   499.600  4053.700   264.100
                  1944   547.500  4379.300   201.600
                  1945   561.200  4840.900   265.000
                  1946   688.100  4900.900   402.200
                  1947   568.900  3526.500   761.500
                  1948   529.200  3254.700   922.400
                  1949   555.100  3700.200  1020.100
                  1950   642.900  3755.600  1099.000
                  1951   755.900  4833.000  1207.700
                  1952   891.200  4924.900  1430.500
                  1953  1304.400  6241.700  1777.300
                  1954  1486.700  5593.600  2226.300
b'US Steel'       1935   209.900  1362.400    53.800
                  1936   355.300  1807.100    50.500
                  1937   469.900  2676.300   118.100
                  1938   262.300  1801.900   260.200
                  1939   230.400  1957.300   312.700
                  1940   361.600  2202.900   254.200
                  1941   472.800  2380.500   261.400
                  1942   445.600  2168.600   298.700
                  1943   361.600  1985.100   301.800
                  1944   288.200  1813.900   279.100
...                          ...       ...       ...
b'Diamond Match'  1945     1.360    65.850     3.310
                  1946     2.240    69.540     3.230
                  1947     3.810    64.970     3.900
                  1948     5.660    68.000     5.380
                  1949     4.210    71.240     7.390
                  1950     3.420    69.050     8.740
                  1951     4.670    83.040     9.070
                  1952     6.000    74.420     9.930
                  1953     6.530    63.510    11.680
                  1954     5.120    58.120    14.330
b'American Steel' 1935     2.938    30.284    52.011
                  1936     5.643    43.909    52.903
                  1937    10.233   107.020    54.499
                  1938     4.046    68.306    59.722
                  1939     3.326    84.164    61.659
                  1940     4.680    69.157    62.243
                  1941     5.732    60.148    63.361
                  1942    12.117    49.332    64.861
                  1943    15.276    75.180    67.953
                  1944     9.275    62.050    69.590
                  1945     9.577    59.152    69.144
                  1946     3.956    68.424    70.269
                  1947     3.834    48.505    71.051
                  1948     5.970    40.507    71.508
                  1949     6.433    39.961    73.827
                  1950     4.770    36.494    75.847
                  1951     6.532    46.082    77.367
                  1952     7.329    57.616    78.631
                  1953     9.020    57.441    80.215
                  1954     6.281    47.165    83.788

[220 rows x 3 columns]

The formula interface for PanelOLS supports the special values EntityEffects and TimeEffects which add entity (fixed) and time effects, respectively.

In [8]:
from linearmodels import PanelOLS
mod = PanelOLS.from_formula('invest ~ value + capital + EntityEffects', data)
res = mod.fit(cov_type='clustered', cluster_entity=True)

In [9]:
res

PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:                 invest   R-squared:                        0.7667
Estimator:                   PanelOLS   R-squared (Between):              0.8223
No. Observations:                 220   R-squared (Within):               0.7667
Date:                Tue, Sep 04 2018   R-squared (Overall):              0.8132
Time:                        20:38:10   Log-likelihood                   -1167.4
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      340.08
Entities:                          11   P-value                           0.0000
Avg Obs:                       20.000   Distribution:                   F(2,207)
Min Obs:                       20.000                                           
Max Obs:                       20.000   F-statistic (robust):             31.490
                                        P-value                           0.0000
Time periods:                      20   Distribution:                   F(2,207)
Avg Obs:                       11.000                                           
Min Obs:                       11.000                                           
Max Obs:                       11.000                                           
                                                                                
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
value          0.1101     0.0144     7.6453     0.0000      0.0817      0.1385
capital        0.3100     0.0500     6.1970     0.0000      0.2114      0.4087
==============================================================================

F-test for Poolability: 49.207
P-value: 0.0000
Distribution: F(10,207)

Included effects: Entity
PanelEffectsResults, id: 0x7f7f3ca061d0



The data set consists of wages and characteristics for men during the 1980s. The entity identifier is nr and the time identified is year. This data is used extensively in Chapter 14 of Introduction to Econometrics by Jeffrey Wooldridge.

Here a MultiIndex DataFrame is used to hold the data in a format that can be understood as a panel. Before setting the index, a year Categorical is created which facilitated making dummies.

In [3]:


from linearmodels.datasets import wage_panel
import pandas as pd
data = wage_panel.load()
year = pd.Categorical(data.year)
data = data.set_index(['nr', 'year'])
data['year'] = year
print(wage_panel.DESCR)
print(data.head())




F. Vella and M. Verbeek (1998), "Whose Wages Do Unions Raise? A Dynamic Model
of Unionism and Wage Rate Determination for Young Men," Journal of Applied
Econometrics 13, 163-183.

nr                       person identifier
year                     1980 to 1987
black                    =1 if black
exper                    labor market experience
hisp                     =1 if Hispanic
hours                    annual hours worked
married                  =1 if married
educ                     years of schooling
union                    =1 if in union
lwage                    log(wage)
expersq                  exper^2
occupation               Occupation code

         black  exper  hisp  hours  married  educ  union     lwage  expersq  \
nr year                                                                       
13 1980      0      1     0   2672        0    14      0  1.197540        1   
   1981      0      2     0   2320        0    14      1  1.853060        4   
   1982      0    

### PooledOLS 

PooledOLS is just plain OLS that understands that various panel data structures. It is useful as a base model. Here the log wage is modeled using all of the variables and time dummies.

In [2]:

from linearmodels.panel import PooledOLS
import statsmodels.api as sm
exog_vars = ['black','hisp','exper','expersq','married', 'educ', 'union', 'year']
exog = sm.add_constant(data[exog_vars])
mod = PooledOLS(data.lwage, exog)
pooled_res = mod.fit()
print(pooled_res)



/home/jon/anaconda3/envs/fastai/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


                          PooledOLS Estimation Summary                          
Dep. Variable:                  lwage   R-squared:                        0.1893
Estimator:                  PooledOLS   R-squared (Between):              0.2066
No. Observations:                4360   R-squared (Within):               0.1692
Date:                Fri, Sep 07 2018   R-squared (Overall):              0.1893
Time:                        22:34:52   Log-likelihood                   -2982.0
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      72.459
Entities:                         545   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                 F(14,4345)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             72.459
                            

### Random effects

The random effects model is virtually identical to the pooled OLS model except that is accounts for the structure of the model and so is more efficient. Random effects uses a quasi-demeaning strategy which subtracts the time average fo the within entity values to account for the common shock.

In [4]:

from linearmodels.panel import RandomEffects
mod = RandomEffects(data.lwage, exog)
re_res = mod.fit()
print(re_res)



                        RandomEffects Estimation Summary                        
Dep. Variable:                  lwage   R-squared:                        0.1806
Estimator:              RandomEffects   R-squared (Between):              0.1853
No. Observations:                4360   R-squared (Within):               0.1799
Date:                Fri, Sep 07 2018   R-squared (Overall):              0.1828
Time:                        22:36:42   Log-likelihood                   -1622.5
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      68.409
Entities:                         545   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                 F(14,4345)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             68.409
                            

The quasi-differencing in the random effects estimator depends on a quantity that depends on the relative variance of the idiosyncratic shock and the common shock. This can be accessed using variance_decomposition.

In [5]:
re_res.variance_decomposition

Effects                   0.106946
Residual                  0.123324
Percent due to Effects    0.464438
Name: Variance Decomposition, dtype: float64

In [6]:
re_res.theta.head()

theta
nr           
13   0.645059
17   0.645059
18   0.645059
45   0.645059
110  0.645059

### The between estimator

The between estimator is an alternative, usually less efficient estimator, can can be used to estimate model parameters. It is particular simple since it first computes the time averages of y
and x

and then runs a simple regression using these averages.

The year dummies are dropped since the averaging removes differences due to the year. expersq was also dropped since it is fairly co-linear with exper. These results are broadly similar to the previous models.

In [7]:
from linearmodels.panel import BetweenOLS
exog_vars = ['black','hisp','exper','married', 'educ', 'union']
exog = sm.add_constant(data[exog_vars])
mod = BetweenOLS(data.lwage, exog)
be_res = mod.fit()
print(be_res)

                         BetweenOLS Estimation Summary                          
Dep. Variable:                  lwage   R-squared:                        0.2155
Estimator:                 BetweenOLS   R-squared (Between):              0.2155
No. Observations:                 545   R-squared (Within):               0.1141
Date:                Fri, Sep 07 2018   R-squared (Overall):              0.1686
Time:                        22:44:55   Log-likelihood                   -194.54
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      24.633
Entities:                         545   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                   F(6,538)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             24.633
                            

### Fixed effects

Entity effects are included by setting entity_effects=True. This is equivalent to including dummies for each entity. In this panel, this would add 545 dummy variables and estimation of the model would be considerably slower. PanelOLS does not actually use dummy variables and instead uses group-wise demeaning to achieve the same effect.
Time-invariant Variables

Time-invariant variables cannot be included when using entity effects since, once demeaned, these will all be 0. Here exper is also excluded since once entity effects and time dummies are incorporated, exper is perfectly co-linear.

In [8]:
from linearmodels.panel import PanelOLS
exog_vars = ['expersq', 'union', 'married', 'year']
exog = sm.add_constant(data[exog_vars])
mod = PanelOLS(data.lwage, exog, entity_effects=True)
fe_res = mod.fit()
print(fe_res)


                          PanelOLS Estimation Summary                           
Dep. Variable:                  lwage   R-squared:                        0.1806
Estimator:                   PanelOLS   R-squared (Between):             -0.0052
No. Observations:                4360   R-squared (Within):               0.1806
Date:                Fri, Sep 07 2018   R-squared (Overall):              0.0807
Time:                        22:48:05   Log-likelihood                   -1324.8
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      83.851
Entities:                         545   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                 F(10,3805)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             83.851
                            

### Time Effects

Time effect can be added using time_effects=True. Here the time dummies are removed. Note that the core coefficients are identical. The only change is in the test statistic for poolability since not the “effects” include both entity and time, whereas before only entity were included.
Effects vs Dummies

For variable which can be consistently estimated, such as time effects in the usual large N, fixed T panel, it isn’t necessary to include these as effects. They can instead be implemented as dummy variables.

In [9]:
from linearmodels.panel import PanelOLS
exog_vars = ['expersq','union','married']
exog = sm.add_constant(data[exog_vars])
mod = PanelOLS(data.lwage, exog, entity_effects=True, time_effects=True)
fe_te_res = mod.fit()
print(fe_te_res)

                          PanelOLS Estimation Summary                           
Dep. Variable:                  lwage   R-squared:                        0.0216
Estimator:                   PanelOLS   R-squared (Between):             -0.0052
No. Observations:                4360   R-squared (Within):              -0.4809
Date:                Fri, Sep 07 2018   R-squared (Overall):             -0.2253
Time:                        22:49:31   Log-likelihood                   -1324.8
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      27.959
Entities:                         545   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                  F(3,3805)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             27.959
                            

https://lectures.quantecon.org/py/pandas_panel.html

